src/compiler/checker.ts

```diff
@@ -18760,7 +18760,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     function instantiateSymbol(symbol: Symbol, mapper: TypeMapper): Symbol {
         const links = getSymbolLinks(symbol);
-        if (links.type && !couldContainTypeVariables(links.type)) {
+        if (links.type && !couldContainTypeVariables(links.type) && !(getDeclarationOfKind(symbol, SyntaxKind.SetAccessor) && couldContainTypeVariables(getWriteTypeOfSymbol(symbol)))) {

```

Comment:
I get the gut feeling that this check is in the wrong place, but I'm not totally sure... This block to me seems like a fast path, e.g. "if there's no type variables then we don't need to do anything because instantiation is a no-op", but this new logic feels tangential.

Indeed if you remove this block entirely, all tests pass (except one tooltip test which I even think is wrong), which would imply that this isn't quite the right place.

But, I'm not totally sure without debugging to see where this is now firing.

Reply:
I did a bit of debugging to understand the problem better.

My understanding is that this block is indeed an optimization, but it has a bug; if the symbol has a setter, then the write type also should be checked for the potential of type variables. Hence why removing this block altogether fixes the issue; it's best effort and it's okay if you don't do it. But it's not okay if you do it and are wrong!

To that end, I don't think that `getWriteTypeOfSymbol` is a good idea here because that may accidentally pull on things; there's a reason the original check wasn't using `getTypeOfSymbol`, for example. But, I do think that there's a version of this that would be more congruent to the existing check which also fixes the bug:

```ts
// If the type of the symbol is already resolved, and if that type could not possibly
// be affected by instantiation, simply return the symbol itself.
if (links.type && !couldContainTypeVariables(links.type)) {
    if (!(symbol.flags & SymbolFlags.SetAccessor)) {
        return symbol;
    }
    // If we're a setter, check writeType.
    if (links.writeType && !couldContainTypeVariables(links.writeType)) {
        return symbol;
    }
}
```

So, basically just the original code, but if it's a setter, we check `writeType` too.

Reply:
Awesome, I didn't realize there was a "writeTypes" in the links.

Reply:
FWIW that PR is #55388 and it seems like it could actually matter for perf (slight memory, maybe one benchmark changed).